scrapy/downloadermiddlewares/redirect.py

```diff
@@ -49,7 +50,7 @@ def _redirect(self, redirected, request, spider, reason):
             redirected.meta['redirect_ttl'] = ttl - 1
             redirected.meta['redirect_urls'] = request.meta.get('redirect_urls', []) + [request.url]
             redirected.meta['redirect_reasons'] = request.meta.get('redirect_reasons', []) + [reason]
-            redirected.dont_filter = request.dont_filter
+            redirected.dont_filter = request.dont_filter or self.fp(request) == self.fp(redirected)
```

Comment:
This does not work for the scenario where A redirects to B which redirects to A. That is why https://github.com/scrapy/scrapy/pull/4314 keeps a set of redirect fingerprints in the request metadata.